Fix PHPStan errors

thelovekesh opened the pull request:
## Summary

- Add `treatPhpDocTypesAsCertain: false` to PHPStan config so that PHPStan doesn't take doc types as certain.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

thelovekesh committed:
Add `treatPhpDocTypesAsCertain: false` config

- This will stop PHPStan to consider docblock type defination as certain

thelovekesh committed:
Remove obsolate `phpstan-ignore-next-line` comment from `MonitorCssTransientCaching::process()`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ampproject/amp-wp?pullRequest=7477) <br/>All committers have signed the CLA.

thelovekesh committed:
Remove additional instanceOf DOMElement checks

thelovekesh committed:
Remove redundant `instanceof WP_Query` check from `amp_is_available()`

thelovekesh committed:
Remove redundant `is_array` check from `AMP_Image_Dimension_Extractor::extract_by_filename_or_filesystem()`

thelovekesh committed:
Refactor to use `new WP_Site_Health()` instead of `WP_Site_Health::get_instance()`

- On AMP Support page it will create a new instance of `WP_Site_Health` anyway. So better to remove dependency on `get_instance()` method.

thelovekesh committed:
Remove additional `is_array` check from `SupportData::process_raw_post_errors`

thelovekesh committed:
Remove additional check to test if current screen is `amp_validated_url` in `AMP_Validated_URL_Post_Type::print_admin_notice()`

thelovekesh committed:
Remove redundant checks for validating if `$action === 'delete'` in `AMP_Validation_Error_Taxonomy::handle_validation_error_update()`

thelovekesh commented:
@westonruter It seems there are a few bugs in the latest PHPStan version so I have started a discussion which later converted into an issue https://github.com/phpstan/phpstan/issues/8980.

thelovekesh commented on src/Support/SupportData.php:
```diff
@@ -247,8 +247,7 @@ public function get_site_info() {
 		if ( class_exists( 'WP_Site_Health' ) ) {
-			$site_health     = method_exists( 'WP_Site_Health', 'get_instance' ) ? WP_Site_Health::get_instance() : new WP_Site_Health();
-			$loopback_status = $site_health->can_perform_loopback();
+			$loopback_status = ( new WP_Site_Health() )->can_perform_loopback();
```
Even though this is wrongly flagged by PHPStan as an error but I think it's in better shape now. On the support page, there will be no instance of `WP_Site_Health` and it will be freshly instantiated anyway.

So IMO it's better to instantiate and read the required data since we will not be using this class instance anywhere else.

westonruter replied:
Yes, this makes sense. And `WP_Site_Health::get_instance()` is the singleton static getter, so it'll instantiate it if it doesn't already exist, and only one copy will ever only ever get constructed.

westonruter replied:
Nevermind, my comment is saying the opposite.

The `WP_Site_Health::get_instance()` method was introduced in WordPress 5.4 via https://github.com/WordPress/wordpress-develop/commit/637b6f545676ef263957263345b032ed14d45744. So that's why we have the `method_exists` check. 

The only potential problem here is that the [constructor](https://developer.wordpress.org/reference/classes/wp_site_health/__construct/) for the class has side effects. But as you say, this class wouldn't have been instantiated on the support page anyway.

westonruter commented on includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php:
```diff
@@ -595,7 +595,7 @@ static function ( $validation_error ) {
 
 		// Remove element if it has illegal CDATA.
-		if ( ! empty( $cdata ) && $node instanceof DOMElement ) {
+		if ( ! empty( $cdata ) ) {
```
Makes sense since we already know it is a `DOMElement`.

westonruter commented on includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php:
```diff
@@ -828,7 +828,7 @@ private function get_required_script_components( DOMElement $node, $tag_spec, $a
 
 		// Check if element needs amp-bind component.
-		if ( $node instanceof DOMElement && ! in_array( 'amp-bind', $this->script_components, true ) ) {
+		if ( ! in_array( 'amp-bind', $this->script_components, true ) ) {
```
Good

westonruter commented on includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php:
```diff
@@ -2334,7 +2334,7 @@ private function has_parent( DOMElement $node, $parent_spec_name ) {
 		// Ensure attributes match; if not move up to the next node.
 		foreach ( $parsed_spec_name['attributes'] as $attr_name => $attr_value ) {
-			if ( $node instanceof DOMElement && strtolower( $node->getAttribute( $attr_name ) ) !== $attr_value ) {
+			if ( strtolower( $node->getAttribute( $attr_name ) ) !== $attr_value ) {
```
Good

westonruter commented on includes/validation/class-amp-validation-error-taxonomy.php:
```diff
@@ -2918,7 +2918,7 @@ public static function handle_validation_error_update( $redirect_to, $action, $t
 		$updated_count = 0;
 		foreach ( $term_ids as $term_id ) {
-			if ( 'delete' === $action && self::delete_empty_term( $term_id ) ) {
+			if ( self::delete_empty_term( $term_id ) ) {
```
Good since the function bails above already if `'delete' !== $action`

westonruter commented on includes/validation/class-amp-validation-error-taxonomy.php:
```diff
@@ -2928,7 +2928,7 @@ public static function handle_validation_error_update( $redirect_to, $action, $t
 
 		$term_ids_count = count( $term_ids );
-		if ( 'edit.php' === $pagenow && 'delete' === $action && 1 === $updated_count ) {
+		if ( 'edit.php' === $pagenow && 1 === $updated_count ) {
```
Good since the function bails above already if `'delete' !== $action`

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -551,7 +551,6 @@ function amp_is_available() {
 		}
 	} elseif ( ! (
 		$queried_object instanceof WP_Post &&
-		$wp_query instanceof WP_Query &&
```
Redundant indeed since above:

```php
if ( ! $wp_query instanceof WP_Query || ! did_action( 'wp' ) ) {
   return;
```

westonruter commented on includes/utils/class-amp-image-dimension-extractor.php:
```diff
@@ -230,7 +230,7 @@ public static function extract_by_filename_or_filesystem( $dimensions ) {
 			}
 
-			if ( ( empty( $image_size ) || ! is_array( $image_size ) ) && file_exists( $image_file ) ) {
+			if ( ( empty( $image_size ) ) && file_exists( $image_file ) ) {
```
Redundant indeed since we check if it is an array above. Nevertheless, this can be further cleaned up a tiny bit more to remove the extra parentheses:

```suggestion
			if ( empty( $image_size ) && file_exists( $image_file ) ) {
```

westonruter commented on includes/validation/class-amp-validated-url-post-type.php:
```diff
@@ -1798,7 +1798,7 @@ public static function print_admin_notice() {
 		// Add notice for PHP fatal error during validation request.
 		$post = get_post();
-		if ( $post instanceof WP_Post && 'post' === get_current_screen()->base && self::POST_TYPE_SLUG === get_current_screen()->post_type ) {
+		if ( $post instanceof WP_Post && 'post' === get_current_screen()->base ) {
```
Indeed redundant since above:

```php
		if ( ! get_current_screen() || self::POST_TYPE_SLUG !== get_current_screen()->post_type ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
			return;
		}
```

westonruter commented on src/Support/SupportData.php:
```diff
@@ -826,7 +825,7 @@ protected function process_raw_post_errors( WP_Post $amp_validated_post ) {
 			$error_source_slugs = array_values( array_unique( $error_source_slugs ) );
 
-			if ( ! empty( $error_source_slugs ) && is_array( $error_source_slugs ) ) {
+			if ( ! empty( $error_source_slugs ) ) {
```
In PHP<8.0, `array_values()` can return `null` if it is not passed an array. Nevertheless, we have ensured above that the `$error_sources` input is an array.

westonruter commented:
> It seems there are a few bugs in the latest PHPStan version so I have started a discussion which later converted into an issue

This is related to these failures, correct?

![image](https://user-images.githubusercontent.com/134745/222281482-263f9820-b867-4844-96f6-e6a804ed0090.png)

thelovekesh committed:
Remove extra parenthesis from if conditional

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh commented:
> This is related to these failures, correct?

Yes. Since we are very specific about these cases we require them as it is, what about ignoring PHPStan on these lines?

thelovekesh committed:
Add temp-baseline.php to be git ignored

thelovekesh committed:
Remove `@phpstan-ignore-next-line` from code files and include them in phpstan baseline instead

thelovekesh committed:
Remove `@phpstan-ignore-next-line` to be appended by amphtml-update.py

thelovekesh committed:
Add `phpstan-baseline.php` to handle PHPStan errors instead of `@phpstan-ignore-next-line` comments

thelovekesh committed:
Add `phpstan-baseline.php` as baseline

thelovekesh commented:
@westonruter I have made a few changes to how we ignore the PHPStan errors, Previously we were using `@phpstan-ignore-next-line` comments which are hard to keep track of, and even remove them when things get fixed upstream.

I have added a new `phpstan-baseline.php` file which will keep track of ignoring required errors with todo. Or we can say a single source to deal with PHPStan errors.

thelovekesh committed:
Fix linting of phpstan-baseline.php

thelovekesh committed:
Update phpstan-baseline.php file doc comment

westonruter commented:
The inclusion of `phpstan-baseline.php` should eliminate the need for `ignoreErrors` in the `phpstan.neon.dist`, correct?

https://github.com/ampproject/amp-wp/blob/963082a52f7ab8ae5c82edfbb70921f987d1d0ff/phpstan.neon.dist#L33-L35

Should that be moved to `phpstan-baseline.php` as well?

westonruter commented on phpstan-baseline.php:
```diff
@@ -0,0 +1,93 @@
+ * 2. Copy the contents of the temp baseline file to this file.
+ *
+ * Please be sure to explain why the error should be ignored in the remark above the error.
+ *
```
```suggestion
 *
 * @codeCoverageIgnore
 *
```

westonruter commented on phpstan-baseline.php:
```diff
@@ -0,0 +1,93 @@
+ * PHPStan Baseline File.
+ * All static analysis errors included in this file will be ignored by PHPStan.
+ *
+ * NOTE: `phpstan analyse —generate-baseline phpstan-baseline.php` should not be used to rewrite this file.
```
```suggestion
 * NOTE: `phpstan analyse --generate-baseline phpstan-baseline.php` should not be used to rewrite this file.
```

thelovekesh committed:
Add `@codeCoverageIgnore` to `phpstan-baseline.php`

Co-authored-by: Weston Ruter <westonruter@google.com>

thelovekesh committed:
Fix command to generate PHPStan baseline

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter commented:
Alternatively, I found that if I run:

```
phpstan analyse --generate-baseline
```

It will generate a `phpstan-baseline.neon` file. The output can then be put straight into the `ignoreErrors` section of `phpstan.neo.dist`. 

Is there a specific reason to use a PHP file for this?

thelovekesh committed:
Move `ignoreErrors` from `phpstan.neon.dist` to `phpstan-baseline.php`

thelovekesh commented:
Yes, but I am not sure if we should use the main config file for this as changes can be there as PHPStan introduces new rules. 

I choose a PHP file for better readability and comments due to better syntax highlighting.

thelovekesh commented:
QA Passed ✅ 

Up until https://github.com/ampproject/amp-wp/commit/4337065a80df965a4d49d2858e6fcced339b90bc, the base branch had no PHPStan issues.
